### PR TITLE
Support for Fisheye Images

### PR DESCRIPTION
**Pull request by canberkgurel**
The existing algorithm doesn't undistort the fisheye images. I added support for the fisheye images.

### CODE REVIEW

File: aruco_detect/src/aruco_detect.cpp [hunk 1]
```diff
@@ -341,7 +342,13 @@ void FiducialsNode::imageCallback(const sensor_msgs::ImageConstPtr & msg) {
         vector <int>  ids;
         vector <vector <Point2f> > corners, rejected;
         vector <Vec3d>  rvecs, tvecs;
-
+        bool isFisheye;
+        nh.getParam("/aruco_detect/isFisheye", isFisheye);
+        if (isFisheye)
+        {
+            cv::fisheye::undistortImage(cv_ptr->image,cv_ptr->image,cameraMatrix,distortionCoeffs,cameraMatrix,cv::Size(cv_ptr->image.cols,cv_ptr->image.rows));
```
Comment (jim-v) [inline aruco_detect/src/aruco_detect.cpp]: If you undistort the image, then you should not pass the distortion coefficients into `estimatePoseSingleMarkers`

**Review by jim-v — COMMENTED** [review]
I'm not the maintainer of this code anymore, so my comments are just suggestions and I have no say in whether this PR will be approved.

The formatting of the code (position of `{` and spaces after `,`) does not match the existing code.

File: aruco_detect/src/aruco_detect.cpp (hunk 1 above)
Comment (canberkgurel) [reply aruco_detect/src/aruco_detect.cpp]: @jim-v thanks for the feedback. I'll fix the formatting. Regarding your comment on `estimatePoseSingleMarkers`, if the user feeds in fisheye images, and passes the distortion coefficients to the `estimatePoseSingleMarkers`, will the aruco detection work accurately, without needing to undistort the images prior to calling the function?

Comment (jim-v) [reply aruco_detect/src/aruco_detect.cpp]: I don't think `estimatePoseSingleMarkers` will do a fisheye undistortion, so you'll either want to undistort the image as your PR does, or undistort the points to get a fisheye undistortion.  In this case you should make sure that the undistortion is not done twice by sending zeros to `estimatePoseSingleMarkers`.

Comment (canberkgurel) [reply aruco_detect/src/aruco_detect.cpp]: @jim-v Thank you! Your comments were very helpful. I will make the updates on the weekend and update my PR accordingly.

**Comment by canberkgurel** [comment]
@rohbotics @jim-v It's been nearly 2 months since I completed this PR, as we wait, there have been changes in master/main branch causing my PR to have merge conflicts. When are you planning to review this?

**Comment by davecrawley** [comment]
@MoffKalast can you please take a look if @rohbotics is unavailable   @canberkgurel I apologize. rohbotics has been out of action for the past few months and is just starting back this week. I've asked another developer to take a look at this - thank you for your contributions.

**Comment by canberkgurel** [comment]
@MoffKalast Any updates on this?

**Comment by MoffKalast** [comment]
@canberkgurel Ah no sorry I've been busy with other things recently. I do have a IMX219-D160 on hand so I could try and test it in the coming days.

**Review by MoffKalast — CHANGES_REQUESTED** [review]
Alright I've tested it out, and while I can't say that the transforms I get are exactly good they're a bit less terrible than with the feature turned off.

There's one thing though, if you do have a camera that's been properly calibrated as a fisheye camera the camera_calibration node (i.e. melodic+ versions) will automatically set the distortion_model to either pinhole or fisheye. Instead of having the extra manual parameter it would make more sense to check the camera_info and get the value from there.

```
header: 
  seq: 1541
  stamp: 
    secs: 1613323317
    nsecs: 663671658
  frame_id: "raspicam"
height: 1232
width: 1648
distortion_model: "fisheye"
```

**Comment by canberkgurel** [comment]
@MoffKalast 

Hi. According to the documentation, the `distortion_model` should be one of the three strings. Please see [sensor_msgs/CameraInfo Message](http://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/CameraInfo.html) and [sensor_msgs/distortion_models.h](http://docs.ros.org/en/melodic/api/sensor_msgs/html/distortion__models_8h_source.html).

- `const std::string PLUMB_BOB = "plumb_bob";`
- `const std::string RATIONAL_POLYNOMIAL = "rational_polynomial";`
- `const std::string EQUIDISTANT = "equidistant";`

Why did you set your `distortion_model` to `fisheye` ? That isn't a standard as far as I can tell from the documentation. Nevertheless, I can still implement what you've suggested. Instead of `fisheye` , I should use `equidistant`.

Also, what did you mean by:

> they're a bit less terrible than with the feature turned off

This sounded more concerning to me than the design choice you have suggested. How did you calibrate your camera? I'm using a factory-calibrated camera, thus I didn't need to do the calibration myself. However, I've calibrated fisheye camera many times in the past. Can you use http://wiki.ros.org/camera_calibration tool in Ubuntu 18.04 & melodic or Ubuntu 20.04 & noetic ?

**Comment by MoffKalast** [comment]
@canberkgurel

I used the melodic branch of camera_calibration  (since the noetic one doesn't compile on Ubuntu16 w/kinetic that I'm using due to the lack of python 3) and it set that parameter by itself.

I've checked and found the piece of code that does it too:

```
def _get_dist_model(dist_params, cam_model):
    # Select dist model
    if CAMERA_MODEL.PINHOLE == cam_model:
        if dist_params.size > 5:
            dist_model = "rational_polynomial"
        else:
            dist_model = "plumb_bob"
    elif CAMERA_MODEL.FISHEYE == cam_model:
        dist_model = "fisheye"
    else:
        dist_model = "unknown"
    return dist_model
```

Comparing it to the noetic branch the "fisheye" param got replaced by "equidistant". I suggest supporting either for better compatibility. After all melodic is not exactly completely end of life and everyone still on kinetic for legacy reasons will be using that as well.

From here: https://github.com/ros-perception/image_pipeline/blob/melodic/camera_calibration/src/camera_calibration/calibrator.py

As for the other thing, it's entirely possible that my calibration wasn't that good, although looking at the rectified image itself it definitely seemed like all the lines were parallel. With these absurd amounts of distortion I wouldn't ever expect the end result to be any good though anyhow, at least not compared to pinhole.

**Comment by canberkgurel** [comment]
@MoffKalast Understood. I'll make the requested modification.

**Comment by canberkgurel** [comment]
@MoffKalast I completed and committed the requested modification. Please check.

**Comment by canberkgurel** [comment]
@MoffKalast Did you test/review the last version? When can it be merged?

**Review by MoffKalast — APPROVED** [review]
I've found some time to retest it and it seems to be working well. Although it does seem to have conflicting code with the main branch, so just resolve that and we can merge.

> This branch has conflicts that must be resolved
Conflicting files
aruco_detect/src/aruco_detect.cpp

**Comment by canberkgurel** [comment]
@MoffKalast I solved the conflicts with the `kinetic-devel` branch. It can be merged now.

**Comment by MoffKalast** [comment]
Okay so it would appear there are still some issues that break on the main branch, especially regarding the estimatePoseSingleMarkers function since its parameter count has changed. I've reverted the merge until that's solved.

**Pull request merged**